lib/value/src/kind/collection.rs

```diff
@@ -0,0 +1,403 @@
+mod exact;
+mod field;
+mod index;
+mod unknown;
+
+use std::collections::BTreeMap;
+
+use super::{merge, Kind};
+use exact::Exact;
+pub use field::Field;
+pub use index::Index;
+use unknown::Unknown;
+
+/// The kinds of a collection (e.g. array or object).
+///
+/// A collection contains one or more kinds for known positions within the collection (e.g. indices
+/// or fields), and contains a global "unknown" state that applies to all unknown paths.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Collection<T: Ord> {
+    known: BTreeMap<T, Kind>,
+
+    /// The kind of other unknown fields.
+    ///
+    /// For example, an array collection might be known to have an "integer" state at the 0th
+    /// index, but it has an unknown length. It is however known that whatever length the array
+    /// has, its values can only be integers or floats, so the `unknown` state is set to those two.
+    unknown: Unknown,
+}
+
+impl<T: Ord> Collection<T> {
+    /// Create a new collection from its parts.
+    #[must_use]
+    pub(super) fn from_parts(known: BTreeMap<T, Kind>, unknown: impl Into<Unknown>) -> Self {
+        Self {
+            known,
+            unknown: unknown.into(),
+        }
+    }
+
+    /// Create a new collection with a defined "unknown fields" value, and no known fields.
+    pub fn unknown(unknown: impl Into<Unknown>) -> Self {
+        Self {
+            known: BTreeMap::default(),
+            unknown: unknown.into(),
+        }
+    }
+
+    /// Create a collection kind of which the encapsulated values can be any kind.
+    #[must_use]
+    pub fn any() -> Self {
+        Self {
+            known: BTreeMap::default(),
+            unknown: Unknown::any(),
+        }
+    }
+
+    /// Create a collection kind of which the encapsulated values can be any JSON-compatible kind.
+    #[must_use]
+    pub fn json() -> Self {
+        Self {
+            known: BTreeMap::default(),
+            unknown: Unknown::json(),
+        }
+    }
+
+    /// Check if the collection fields can be of any kind.
+    ///
+    /// This returns `false` if at least _one_ field kind is known.
+    #[must_use]
+    pub fn is_any(&self) -> bool {
+        self.known.iter().all(|(_, k)| k.is_any()) && self.unknown.is_any()
+    }
+
+    /// Get the "known" and "unknown" parts of the collection.
+    #[must_use]
+    pub(super) fn into_parts(self) -> (BTreeMap<T, Kind>, Unknown) {
+        (self.known, self.unknown)
+    }
+
+    /// Get the "known" field value kinds.
+    #[must_use]
+    pub fn known(&self) -> &BTreeMap<T, Kind> {
+        &self.known
+    }
+
+    /// Get a mutable reference to "known" field value kinds.
+    pub fn known_mut(&mut self) -> &mut BTreeMap<T, Kind> {
```

Comment:
I think Clippy doesn't warn here because we take `&mut self`, which often (but not always, such as here) implies the method has side effects, that don't always require the use of the return value.